### PR TITLE
QA: Merge develop

### PR DESCRIPTION
**Pull request by btcdrak**
ping @reorder please review consensus critical merge.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -2371,6 +2399,92 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
     return true;
 }
 
+bool ContextualCheckBlockHeader(const CBlockHeader& block, CValidationState& state, CBlockIndex * const pindexPrev)
```
Comment (btcdrak) [inline src/main.cpp:2402]: Please review this diff block.

File: src/main.cpp [hunk 2]
```diff
@@ -2395,62 +2509,16 @@ bool AcceptBlockHeader(const CBlockHeader& block, CValidationState& state, CBloc
 
     // Get prev block index
     CBlockIndex* pindexPrev = NULL;
-    int nHeight = 0;
     if (hash != Params().HashGenesisBlock()) {
         BlockMap::iterator mi = mapBlockIndex.find(block.hashPrevBlock);
         if (mi == mapBlockIndex.end())
             return state.DoS(10, error("%s : prev block not found", __func__), 0, "bad-prevblk");
         pindexPrev = (*mi).second;
-        nHeight = pindexPrev->nHeight+1;
```
Comment (btcdrak) [inline src/main.cpp:2404]: This code was moved above.

File: src/main.cpp [hunk 3]
```diff
@@ -2395,62 +2509,16 @@ bool AcceptBlockHeader(const CBlockHeader& block, CValidationState& state, CBloc
...
-        if (!Checkpoints::CheckBlock(nHeight, hash))
-            return state.DoS(100, error("%s : rejected by checkpoint lock-in at %d", __func__, nHeight),
-                             REJECT_CHECKPOINT, "checkpoint mismatch");
-
-        // Don't accept any forks from the main chain prior to last checkpoint
-        CBlockIndex* pcheckpoint = Checkpoints::GetLastCheckpoint();
-        if (pcheckpoint && nHeight < pcheckpoint->nHeight)
-            return state.DoS(100, error("%s : forked chain older than last checkpoint (height %d)", __func__, nHeight));
-
-        if ((Params().AllowMinDifficultyBlocks() && nHeight < 300000) || nHeight < 451000) {
```
Comment (btcdrak) [inline src/main.cpp:2427]: Particularly this stuff.

**Comment by reorder** [comment]
Basically, just CheckBlockHeader split to context-dependent and context-independent parts, looks good. ACK.

**Pull request merged**